Project 11 [Стразов, Зураев]

BulatZuraev opened the pull request.

BulatZuraev committed:
add project

ARostovsky commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+
+-- Функции объектов
+CREATE TYPE function AS ENUM ('Жилое помещение', 'Кафе', 'Спорт. площадка', 'Бассейн', 'Штаб делегации', 'Столовая');
```
I'd rather call it object_type, but it's more a nitpick.

BulatZuraev replied:
done

ARostovsky commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+CREATE TABLE delegation(
+    id SERIAL PRIMARY KEY,
+    -- У одной страны только одна национальная делегация
+	country TEXT NOT NULL UNIQUE,
```
Probably it's worth having an ENUM for countries.

BulatZuraev replied:
done

ARostovsky commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+    id SERIAL PRIMARY KEY,
+    -- У одной страны только одна национальная делегация
+	country TEXT NOT NULL UNIQUE,
+    -- В одном объекте может жить только одна делегация
```
I didn't find such condition, but `Кроме того, в одном из объектов должен быть штаб делегации`. Seems like `NOT NULL` is enough here.

BulatZuraev replied:
done

ARostovsky commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+-- Отношение между объектами и
+-- видами спорта для которых они предназначены
+-- отношение inf <-> inf
+CREATE TABLE object_ToSport(
```
So do you prefer CamelCase or snake_case after all?)

dbarashev replied:
_diversity!_

BulatZuraev replied:
done

ARostovsky commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+-- Задание содержит id волонтера, дату выполнения и текстовое описание.
+CREATE TABLE VolunteerTask(
+    id SERIAL,
+	volunteer_id TEXT NOT NULL,
```
Why do you prefer text foreign key over having integer id field in volunteer table and using it as a foreign key?

BulatZuraev replied:
done

ARostovsky reviewed (commented):
In general, I like your project, good job guys!

Ayokontol commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+    id SERIAL, 
+    address TEXT NOT NULL,
+    function FUNCTION NOT NULL, 
+    name TEXT NOT NULL,
```
Имя объекта опционально (у жилого дома не будет названия), поэтому NOT NULL лишнее.

BulatZuraev replied:
done

Ayokontol reviewed (commented):
Все круто

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+CREATE TABLE delegation(
+    id SERIAL PRIMARY KEY,
+    -- У одной страны только одна национальная делегация
+	country TEXT NOT NULL UNIQUE,
```
Будете ли вы счастливы, если какой-то безвестный клерк вобьет в этот столбец значение "Rusia" или "sdfsdfasdf"?

BulatZuraev replied:
done

BulatZuraev replied:
Думаю дефект появился на первой итерации.

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+	house_id INT NOT NULL UNIQUE,
+	name TEXT,
+    -- У делегации только один руководитель
+	leader_id INT NOT NULL UNIQUE,
```
что это за id? Любое целое число сюда вможно вбить, лишь бы только повторов не было?

BulatZuraev replied:
done

BulatZuraev replied:
Думаю дефект появился на второй итерации.

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+    --card_id может принадлежать только одному спортсмену
+	PRIMARY KEY(card_id),
+	-- Спортсмен может жить только в одном объекте.
+	UNIQUE(card_id, house_id),
```
Правда ли, что если card_id уникален сам по себе, то добавление ему в пару еще какого-то атрибута более уникальным его не сделает?

BulatZuraev replied:
done

BulatZuraev replied:
Дефект появился на третьей итерации, его причиной стало временное помешательство (невнимательность).

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+	FOREIGN KEY(house_id) REFERENCES object_(id),
+    -- Проверка на возраст, вес и рост.
+	CHECK (age > 0), 
+    CHECK (weight > 0), 
```
Это граммы, килограммы, тонны? Какая единица измерения?

BulatZuraev replied:
done

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+
+--------------------------------------------- SPORTS ----------------------------------------------------------------
+-- Служебная табличка с id видов спорта
+CREATE TYPE sport AS ENUM('Фехтование', 'Баскетбол', 'Плавание', 'Футбол');
```
Что будете делать, когда надо будет добавить еще один вид спорта?

BulatZuraev replied:
done

BulatZuraev replied:
Думаю дефект появился на первой итерации.

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+-- видами спорта для которых они предназначены
+-- отношение inf <-> inf
+CREATE TABLE object_ToSport(
+    object__id INT NOT NULL, 
```
два подчеркивания, найс. Долой рутину из написания SQL запросов! :)

BulatZuraev replied:
done

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+    -- С заданным именем может быть только одно соревнование
+    name TEXT NOT NULL UNIQUE,
+	object__id INT NOT NULL,
+    date TIMESTAMP NOT NULL, 
```
С точностью до микросекунды!

BulatZuraev replied:
done

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+CREATE TYPE medal_type AS ENUM('Золото', 'Серебро', 'Бронза');
+
+-- Записи о выигранных медалях
+CREATE TABLE medal(
```
Записи в этой таблице как-нибудь коррелируют с записями в таблице  SportsmanToCompetition? Можно ли получить медаль, не участвуя в соревновании?

BulatZuraev replied:
Думаю дефект появился на первой или второй итерации, для уточнения призываю @IvanStrazov

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+-- Волонтер определяется своей id картой и при этом волонтеры должны иметь уникальные телефоны.
+-- Может быть неконсистентность, что id волонтера совпадает с id спортсмена.
+CREATE TABLE volunteer(
+    card_id TEXT NOT NULL,
```
тут TEXT а у спортсмена varchar(5). Разные люди писали код?

BulatZuraev replied:
done

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+
+-- Отношение 1 <-> inf между волонтерами и спортсменами
+
+CREATE TABLE VolunteerToSportsman(
```
Возможно, задачу назначения не более одного волонтера спортсмену можно решить попроще, без создания ново таблицы?

BulatZuraev replied:
done

dbarashev commented on project11/project11.sql:
```diff
@@ -0,0 +1,248 @@
+    id SERIAL,
+	volunteer_id TEXT NOT NULL,
+	task_description TEXT NOT NULL,
+	date TIMESTAMP NOT NULL,
```
микросекунды, опять же.

BulatZuraev replied:
done

dbarashev commented:
@BulatZuraev @IvanStrazov отметил эмоджей с глазками комментарии к недостаткам, причину и итерацию появления которых нужно выяснить

BulatZuraev committed:
fixes

rename function enum to object_type
add country enum
remove house_id unique constraint at delegation table
replace camelcase with snakecase
remove name not null constraint at object_ table
add leader table
remove redundant unique constraints at sportsman table
add age and weight units at sportsman table description
replace sport enum with table
replace timestamp  with date and time(0)
move medal info to sportsman_to_competition table
change card_id type from text to varchar(5) at volunteer table
move volunteer to sportsman relation to sportsman table

BulatZuraev committed:
add id for sportsman and volunteer tables

SokovIlyaArchiv committed:
Completed

marinskiy committed:
03122020 progress
